JavaScript/testing.md

```diff
@@ -0,0 +1,109 @@
+
+**Types** establish explicit contractual obligations between functions. If something is a `string` or `undefined`, it is known to both the developer and to the type system. If we try to do something dangerous when that `string` is possibly `undefined`, the type system will alert us, sometimes even preventing the code from compiling. This forces us to code defensively at every step of our application, preventing us from trying to manipulate data in a way that will cause errors.
+
+It also forces us to think more critically about the data and functions we create. Eventually, making sure our code is defensive will become second nature since we will learn how to avoid all sorts of type errors. It will prevent spelling errors and errors like trying to access a variable of a parent object that is possibly `undefined`. It will also aid heavily in ensuring that our reducers are functioning as expected, since we can type reducers to always return a certain type of state - `HomeState` for example. Anything that is returned within that reducer will need to fit that existing type.
+
+If any code is refactored, it will immediately point out errors with accessing incorrect or stale data. For example, if we were to refactor a slice of our state, any `mapStateToProps` function that pull from it will be aware of the new state and error if we are trying to pull state that does not exist. Similarly, any `prop` we pass will now have an updated signature, preventing any runtime errors, like converting an array into an object and trying to map over that object. Refactors like this are frequent once a project starts reaching a few dozen components.
+
+<a name="testing-for-accidental-code-changes"></a>
+
+### Testing for Accidental Code Changes
+
+The two main defenses against accidental code changes are **Code Reviews** and [**Snapshot Tests**](https://jestjs.io/docs/en/snapshot-testing). Code Reviewers can spot accidental changes, which is they are so important. **Snapshot Tests** can also point out to the developer that their code has changed in a file or module that they were not supposed to be editing. For example, if we change some of the login logic and a snapshot test fails in our search component, we've probably done something wrong or did a code change where we shouldn't have.
+
+<a name="testing-in-summary"></a>
+
+## Testing in Summary
+
+Testing is always a tradeoff between time, money, and coverage. Using the above methods, we can ensure that we'll test for the majority of use cases that cause bugs to arrive. In the future, we will explore more E2E techniques, which can act as a safeguard for making sure that flows function correctly. As a heuristic, test coverage percentages should be taken with a grain of salt. It causes us to write feel-good tests, like simple setters in our reducers, and lowers our velocity. Aiming for 100% code coverage is too idealistic and will ensure false confidence in our codebase.
```

Comment:
```suggestion
Testing is always a tradeoff between time, money, and coverage. Using the above methods, we can ensure that we'll test for the majority of use cases that cause bugs to arrive. In the future, we will explore more end-to-end (E2E) techniques, which can act as a safeguard for making sure that flows function correctly. As a heuristic, test coverage percentages should be taken with a grain of salt. It causes us to write feel-good tests, like simple setters in our reducers, and lowers our velocity. Aiming for 100% code coverage is too idealistic and will ensure false confidence in our codebase.
```

Reply:
I'd say leave off the part about code coverage, since it primed me into thinking there was even more material below. Additionally, the value of code coverage in general depends on how it is being leveraged. In my opinion, the ideal use involves having most of our tests perform tasks E2E that would cover most of the code branches. Then more specific tests could be written to satisfy code coverage and ensure that nothing breaks in runtime under any condition. In that case, a code coverage of 100% would be a great guarantee of code quality.

Reply:
I do agree with you that testing simple Redux setters is not necessary, since it's likely that it can be covered with a good E2E test.

Reply:
will remove the part about code coverage. definitely agree about ideally having E2E tests that cover most of our branches 👍

I think I would disagree that code coverage = code quality though. If we were testing a single method that accepted a range of inputs and transformed that data, we could theoretically just pass in a single pair of numbers that satisfy the conditions and run through the possible branches. We would get 100% code coverage, but there's still the possibility we didn't defensively code and a divide by zero error could pop up. I think this would then have the opposite effect, where the developer will then be a bit overly confident in their code when in reality there are still problems.

I feel like there's still a lot of user error in tests as well (I will say that it definitely exists in creating types as well, so not just trying to disparage tests 😆). I've come across numerous tests that just describe what the code itself is doing, when the code is wrong. I've also seen people just modify their test case to match their output since their own smoke testing passed.

I do suppose that those extra tests to reach 100% would really be contextual and driven by the project and code itself. Know we're speaking about this very abstractly and i'm not trying to just argue over semantics, but I think each time we strive for that, we should compare the time / cost to implement and its impact on delivering as well. Know you were speaking idealistically about reaching 100% _after_ E2E tests, but I think this does always play into it

For a project like Westfield where we have a lot of time and resources, I would be inclined to agree, but for something like Skillzilla where we were under a tight deadline, the probability that additional tests like that would reduce bugs would be low imo.  I feel like going past 80% code coverage is when the positive benefits relative to the time to implement tend to exponentially decrease, especially as maintaining that extra 20% will always be a time sink moving forward, especially if those 80% test cases are what's run in production 95% of the time compared to the other 20%.